Git module refactor and new git clean files feature

raags opened the pull request:
There are three commits with each one doing the following :
1. Simple cleanup of trailing space and minor fixes
2. Refactored code to remove unused vars and renamed some to make them explicit. Changed the exit json structure (but still backword compatible), which I believe is easier to work with.
3. Finally added clean_untracked and clean_ignored files option, which I had opened for discussion here : https://github.com/ansible/ansible-modules-core/issues/2656

I initially just wanted to add the clean files feature, but it was cumbersome to do that without making these additional changes. I've run the integration tests against these changes, and I've also added more, which I'll push to the ansible repository now.

raags committed:
Make changed status explicit and exit json consistent

* The changed variable is explicitly set whenever a change occurs, this
was not clear earlier.
* The repo_updated var was a left over and being used for another purpose
altogether. Changed that to update_repo now, which is clearer.
* Module exits at a single point now with unified json output that will
always have same strucuture. Check mode should also do the same.

raags committed:
New option to clean untracked and ignored files

* clean_ignored and clean_untracked options will clean ignored
and untracked file respectively.
* Added untracked_files and ignored_files keys to the output, and
changed local_mods key to local_mod_files to keep it consistent.

raags commented:
The refactoring touches code from the following commits that I got from git blame, might be helpful in reviewing the changes :
https://github.com/raags/ansible-modules-core/commit/050a462b2a835aa87a973ae1e229ba227fc24e54
https://github.com/raags/ansible-modules-core/commit/633438bfbb097ba0b05c74b31f0cbcaa690fb8a0
https://github.com/raags/ansible-modules-core/commit/b77bf7a9aa1889fbee75f3db17c89816ca7c7838
https://github.com/raags/ansible-modules-core/commit/c6f4d4328bb341d13089cc6374bae798699fadf5

gregdek commented:
Thanks @raags for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

gregdek commented:
Thanks @raags for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

raags commented:
ready_for_review

gregdek commented:
Thanks @raags for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
Thanks @raags for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

raags commented:
ready_for_review @abadger @gregdek can someone from core team please review this?

gregdek commented:
Thanks @raags for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

bcoca commented on source_control/git.py:
```diff
@@ -103,6 +103,20 @@
+        choices: [ "yes", "no" ]
+        version_added: "2.1"
+        description:
+            - If C(yes), clean untracked files and directories in the repository
```
I think a cleaner interface is an option `clean: None|ignored|untracked|all`

raags replied:
Yea, that looks much cleaner, I'll make the changes.

bcoca commented:
there are a bunch of unrelated style changes, this is not OK as it changes attribution for no good reason.

raags commented:
@bcoca imho trailing spaces shouldn't be accepted, its sloppy. I removed dead code, and changed a non  pythonic check that pep complained about. I kept it a separate commit which makes it clear its minor formatting change.

note: closed the pr by mistake, and reopened it which remoed the needs_revision tag

raags commented:
Bump @bcoca I've made the suggested changes, let me know if anything else is required.

bcoca commented:
still have a bunch of unrelated style changes...

raags commented:
@bcoca I've removed the formatting changes, please review.

ready_for_review

gregdek commented:
Thanks @raags for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

robinro commented:
@raags why did you put the `clean_files` call after the fetch/switch_version part?
`switch_version` will fail if there are locally modified files and cleaning them before would be similar to `force` to work around that (but more specific). 

Independent of when clean is run an integration test of `clean` in combination with a `version` change would be nice.

raags commented:
@robinro `clean` is for ignored and untracked files; These will not interfere with fetching or changing versions. The `force` option deals with what you are referring to, which will behave as it is currently.

robinro commented:
cleaning can interfere with switch_version if the file is not tracked in the current branch, but tracked in the other branch; if cleaned first, the version switch will work without force, but break if clean is run afterwards

gregdek commented:
Thanks @raags for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@raags A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@raags Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@raags A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@raags Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@raags A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@raags Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@raags A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@raags Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@raags A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@raags Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@raags A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]
